sources/httpUtils.ts

```diff
@@ -30,60 +26,55 @@ export async function fetchUrlStream(url: string, options: RequestOptions = {})
     }
   }
 
-  return new Promise<IncomingMessage>((resolve, reject) => {
-    const createRequest = (url: string) => {
-      const request: ClientRequest = https.get(url, {...options, agent: proxyAgent}, response => {
-        const statusCode = response.statusCode;
-
-        if ([301, 302, 307, 308].includes(statusCode as number) && response.headers.location)
-          return createRequest(response.headers.location as string);
-
-        if (statusCode != null && statusCode >= 200 && statusCode < 300)
-          return resolve(response);
-
-        return reject(new Error(`Server answered with HTTP ${statusCode} when performing the request to ${url}; for troubleshooting help, see https://github.com/nodejs/corepack#troubleshooting`));
-      });
+  let response;
+  try {
+    response = await globalThis.fetch(input, {
+      ...init,
+      dispatcher: agent,
+    });
+  } catch (error) {
+    throw new Error(
+      `Error when performing the request to ${input}; for troubleshooting help, see https://github.com/nodejs/corepack#troubleshooting`,
+      {cause: error},
+    );
+  }
 
-      request.on(`error`, err => {
-        reject(new Error(`Error when performing the request to ${url}; for troubleshooting help, see https://github.com/nodejs/corepack#troubleshooting`));
-      });
-    };
+  if (!response.ok) {
+    await response.arrayBuffer();
+    throw new Error(
+      `Server answered with HTTP ${response.status} when performing the request to ${input}; for troubleshooting help, see https://github.com/nodejs/corepack#troubleshooting`,
+    );
+  }
 
-    createRequest(url);
-  });
+  return response;
 }
 
-export async function fetchAsBuffer(url: string, options?: RequestOptions) {
-  const response = await fetchUrlStream(url, options);
-
-  return new Promise<Buffer>((resolve, reject) => {
-    const chunks: Array<Buffer> = [];
+export async function fetchAsJson(input: string | URL, init?: RequestInit) {
+  const response = await fetch(input, init);
+  const json: any = await response.json();
+  return json;
+}
 
-    response.on(`data`, chunk => {
-      chunks.push(chunk);
-    });
+export async function fetchUrlStream(input: string | URL, init?: RequestInit) {
+  const response = await fetch(input, init);
+  const webStream = response.body;
+  assert(webStream, `Expected stream to be set`);
+  const stream = Readable.fromWeb(webStream);
+  return stream;
+}
 
-    response.on(`error`, error => {
-      reject(error);
-    });
+async function getProxyAgent(input: string | URL) {
+  const {getProxyForUrl} = await import(`proxy-from-env`);
 
-    response.on(`end`, () => {
-      resolve(Buffer.concat(chunks));
-    });
-  });
-}
+  const proxy = getProxyForUrl(input.toString());
```

Comment:
Looking at the implementation, it looks like converting it to a string a waste: https://github.com/Rob--W/proxy-from-env/blob/96d01f8fcfdccfb776735751132930bbf79c4a3a/index.js#L20-L25
Any reason not to pass the input as is?
```suggestion
  const proxy = getProxyForUrl(input);
```

Reply:
```
Argument of type 'string | URL' is not assignable to parameter of type 'string | Url'.
  Type 'URL' is not assignable to type 'string | Url'.
    Type 'URL' is missing the following properties from type 'Url': auth, path, slashes, query ts(2345)
```

Reply:
Maybe we can patch the types as the look wrong? The code only needs `protocol`, `host`, and `port` – which exists on WHATWG URLs

Reply:
It would be simpler to ignore the error in that case.
```suggestion
  // @ts-expect-error - The internal implementation is compatible with a WHATWG URL instance
  const proxy = getProxyForUrl(input);
```

Reply:
SGTM. For reference, I've opened https://github.com/Rob--W/proxy-from-env/pull/22 to try to address this upstream.